H.Ws/H.W_4.py

```diff
@@ -0,0 +1,236 @@
+
+    fig_no_relationship = go.Figure(
+        data=go.Heatmap(z=conf_matrix, zmin=0, zmax=conf_matrix.max())
+    )
+    fig_no_relationship.update_layout(
+        title="Categorical Predictor by Categorical Response (with relationship)",
+        xaxis_title="Response",
+        yaxis_title="Predictor",
+    )
+    fig_no_relationship.show()
+    pass
+
+
+if resp_test == "boolean" and pred_test == "categorical":
+    for i in x and y:
+        cat_response_cat_predictor()
+
+
+# 2_cat response - cont predictor
+def cat_resp_cont_predictor():
+    n = 200
+    # test matrix
+    # x = titanic_df['age'].to_numpy()
+    # x = x[~numpy.isnan(x)]
+    # y = titanic_df['fare'].to_numpy()
+    # y = y[~numpy.isnan(y)]
+    # z = [x, y]
+
+    group_labels = ["Response"]
+
+    # Create distribution plot with custom bin_size
+    fig_1 = ff.create_distplot([x[val]], group_labels, bin_size=0.2)
+    fig_1.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title="Predictor",
+        yaxis_title="Distribution",
+    )
+    fig_1.show()
+
+    fig_2 = go.Figure()
+    for y, curr_group in zip([x[val]], group_labels):
+        fig_2.add_trace(
+            go.Violin(
+                x=numpy.repeat(curr_group, n),
+                y=y,
+                name=curr_group,
+                box_visible=True,
+                meanline_visible=True,
+            )
+        )
+    fig_2.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title="Response",
+        yaxis_title="Predictor",
+    )
+    fig_2.show()
+    pass
+
+
+if resp_test == "boolean" and pred_test == "continuous":
+    for i in x:
+        cat_resp_cont_predictor()
+
+
+# 3_con response - cat predictor
+def cont_resp_cat_predictor():
+    n = 200
+    # Add histogram data
+    # a = titanic_df['sex']
+    # b = titanic_df['pclass']
+    # c = titanic_df['deck']
+    # z = [a, b, c]
+
+    group_labels = ["Group 1", "Group 2", "Group 3"]
+
+    # Create distribution plot with custom bin_size
+    fig_1 = ff.create_distplot([x[val]], group_labels, bin_size=0.2)
+    fig_1.update_layout(
+        title="Continuous Response by Categorical Predictor",
+        xaxis_title="Response",
+        yaxis_title="Distribution",
+    )
+    fig_1.show()
+
+    fig_2 = go.Figure()
+    for y, curr_group in zip([x[val]], group_labels):
+        fig_2.add_trace(
+            go.Violin(
+                x=numpy.repeat(curr_group, n),
+                y=y,
+                name=curr_group,
+                box_visible=True,
+                meanline_visible=True,
+            )
+        )
+    fig_2.update_layout(
+        title="Continuous Response by Categorical Predictor",
+        xaxis_title="Groupings",
+        yaxis_title="Response",
+    )
+    fig_2.show()
+    pass
+
+
+if resp_test == "continuous" and pred_test == "categorical":
+    for i in x:
+        cont_resp_cat_predictor()
```

Comment:
I see you are defining a function and calling it right away...I would suggest use a main function to read the data and indentify the type of predicors and then call these functions according. 
This will provide better visibility and it would ensure that your code is working correct.
Also easier to debug...